node/network/collator-protocol/src/collator_side/mod.rs

```diff
@@ -868,38 +947,78 @@ async fn handle_incoming_request<Context>(
 		Some(our_para_id) => {
 			gum::warn!(
 				target: LOG_TARGET,
-				for_para_id = %req.payload.para_id,
+				for_para_id = %para_id,
 				our_para_id = %our_para_id,
 				"received a `CollationFetchingRequest` for unexpected para_id",
 			);
 		},
 		None => {
 			gum::warn!(
 				target: LOG_TARGET,
-				for_para_id = %req.payload.para_id,
+				for_para_id = %para_id,
 				"received a `RequestCollation` while not collating on any para",
 			);
 		},
 	}
 	Ok(())
 }
 
-/// Our view has changed.
+/// Peer's view has changed. Send advertisements for new relay parents
+/// if there're any.
 #[overseer::contextbounds(CollatorProtocol, prefix = self::overseer)]
 async fn handle_peer_view_change<Context>(
 	ctx: &mut Context,
 	state: &mut State,
 	peer_id: PeerId,
 	view: View,
 ) {
-	let current = state.peer_views.entry(peer_id.clone()).or_default();
+	let PeerData { view: current, version } = match state.peer_data.get_mut(&peer_id) {
+		Some(peer_data) => peer_data,
+		None => return,
+	};
 
 	let added: Vec<Hash> = view.difference(&*current).cloned().collect();
 
 	*current = view;
 
 	for added in added.into_iter() {
-		advertise_collation(ctx, state, added, peer_id.clone()).await;
+		let block_hashes = match state
+			.per_relay_parent
+			.get(&added)
+			.map(|per_relay_parent| per_relay_parent.prospective_parachains_mode)
+		{
+			Some(ProspectiveParachainsMode::Disabled) => std::slice::from_ref(&added),
+			Some(ProspectiveParachainsMode::Enabled) => state
+				.implicit_view
+				.known_allowed_relay_parents_under(&added, state.collating_on)
+				.unwrap_or_default(),
+			None => {
+				gum::trace!(
+					target: LOG_TARGET,
+					?peer_id,
+					new_leaf = ?added,
+					"New leaf in peer's view is unknown",
+				);
+				continue
+			},
+		};
+
+		for block_hash in block_hashes {
+			let per_relay_parent = match state.per_relay_parent.get_mut(block_hash) {
+				Some(per_relay_parent) => per_relay_parent,
+				None => continue,
+			};
+			advertise_collation(
```

Comment:
Reviewed that this is only advertising to peers with the relay-parent in their implicit view.